v1.0/pvp-specification.md

```diff
@@ -48,7 +48,10 @@ ordering of the components. For example, 2.0.1 \> 1.3.2, and 2.0.1.0 \> 2.0.1.
 version number. When a package is updated, the following rules govern
 how the version number must change relative to the previous version:
 
-1. *Breaking change*. If any entity was removed, or the types of any entities
+1. *Internal modules*. If a change only touches entities defined in modules
+whose names include the string `.Internal` then *A.B* **MAY** remain unchanged.
```

Comment:
I feel this is somewhat ambiguous. 

1. what about `Foo.Internal.Bar`?
2. what about `Foo.InternalBar`?

Both include the string `.Internal`.

Reply:
Yes, perhaps appealing to the module hierarchy instead of modules names as strings would be more appropriate here.

Reply:
What if a package is called `Internal`? Do we just say "don't do that then", or should it be a non-initial component?

The way I see it, we would want both `AwesomeLibrary.Datatypes.PriorityQueue.Internal` and `GHC.Internal.ByteMunging` to be considered internals, but `Internal.Datatypes.List` to not be.

Perhaps it's OK to just lose that namespace, though - it would be pretty confusing.

Reply:
That's exactly what's problematic with magic names.

PVP already implicitly makes a connection to cabal file syntax. So what if we we add support for "internal" modules to cabal first and then adjust PVP to reference it.

E.g.

- `exposed-modules` (exposed and PVU compliant)
- `other-modules` (not exposed)
- `internal-modules` (exposed, but exempt from PVP)

This also allows easy static analysis and doesn't depend on module names.

It could be argued PVP is more general than haskell, but reading the spec carefully, that doesn't seem to be the case anyway.

The spec could also leave this open and intentionally vague and just point out that breaking changes can be exempted by a language's metadata format, specifying those modules in an appropriate format.

Reply:
> So what if we we add support for "internal" modules to cabal first and then adjust PVP to reference it.

This seems to me to be a great thing to have done years ago. But today, we happen to have a spec that says to do one thing, and a lot of people who actually do something not quite what the spec says. Propagating a new file format to do this means convincing everyone to abandon their current practices, and move to something else, as well as waiting for the new versions of the various tools to propagate and spread.

Why not formalize what people actually do, which seems to work reasonably well in practice today?

Reply:
> Why not formalize what people actually do, which seems to work reasonably well in practice today?

Because they don't. People treat `.Internal` modules of `bytestring` or `text` as stable and under PVP. Please address that specifically.

Reply:
> Why not formalize what people actually do, which seems to work reasonably well in practice today?

Because that's how specs generally deteriorate and simply become documentation of current (good and bad) practices.

I'd rather work towards making the spec more general and leave space for languages and tooling to decide on non-critical things.

I also don't see any particular time pressure here. Are we in a rush?